File: Python/494.py
class Solution(object):
    def findTargetSumWays(self, nums, S):
        """
        :type nums: List[int]
        :type S: int
        :rtype: int
        """
        ss = sum(nums)
        if ss < S:
            return 0
        if (ss + S)%2 != 0:
            return 0
        target = (ss+S)/2
        dp = [0]*(target+1)
        dp[0] = 1
        for n in nums:
            for i in range(target, n-1, -1):
                dp[i] += dp[i-n]
        return dp[target]

'''
iterate dp backwards, cause dp is reduced
from two dimensions to one dimension.
such as dp[i] = dp[i] + dp[i-n]
right value "dp[i], dp[i-n]" is the dp value last time it computed.
iterate forwards would pollute the old dp value, thus the result will be duplicated.
'''

File: Cpp/494.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int S) {
        int sum=0;
        for(auto&& n:nums)
            sum+=n;
        if(sum<S||(S+sum)&1)
            return 0;
        int s1=(S+sum)/2;
        vector<int> dp(s1+1,0);
        dp[0]=1;
        for(auto&& n:nums){
            for(int i=s1;i>=n;--i){
                //as for dp[i], if not pick n, it's dp[i], otherwise it's dp[i-n], thus dp[i]=dp[i]+dp[i-n]
                dp[i]+=dp[i-n];
            }
        }
        return dp[s1];
    }
};

/*
 
 s1-s2=S
 s1=S+s2
 2*s1=S+s1+s2
 s1=1/2(S+sum)
 
 s1 must be even.
 
 
 */

int main(){

}
